packages/@aws-cdk/pipelines/lib/codepipeline/codepipeline-engine.ts

```diff
@@ -0,0 +1,888 @@
+import { BuildDeploymentOptions, IDeploymentEngine } from '../main/engine';
+import { appOf, assemblyBuilderOf, embeddedAsmPath, obtainScope } from '../private/construct-internals';
+import { toPosixPath } from '../private/fs';
+import { enumerate, flatten, maybeSuffix } from '../private/javascript';
+import { writeTemplateConfiguration } from '../private/template-configuration';
+import { CodeBuildFactory, mergeCodeBuildOptions, stackVariableNamespace } from './_codebuild-factory';
+import { ArtifactMap } from './artifact-map';
+import { CodeBuildStep } from './codebuild-step';
+import { CodePipelineActionFactoryResult, ICodePipelineActionFactory } from './codepipeline-action-factory';
+
+/**
+ * Creation properties for a `CodePipelineEngine`
+ */
+export interface CodePipelineEngineProps {
+  /**
+   * The name of the CodePipeline pipeline
+   *
+   * @default - Automatically generated
+   */
+  readonly pipelineName?: string;
+
+  /**
+   * Create KMS keys for the artifact buckets, allowing cross-account deployments
+   *
+   * The artifact buckets have to be encrypted to support deploying CDK apps to
+   * another account, so if you want to do that or want to have your artifact
+   * buckets encrypted, be sure to set this value to `true`.
+   *
+   * Be aware there is a cost associated with maintaining the KMS keys.
+   *
+   * @default false
+   */
+  readonly crossAccountKeys?: boolean;
+
+  /**
+   * CDK CLI version to use in self-mutation and asset publishing steps
+   *
+   * If you want to lock the CDK CLI version used in the pipeline, by steps
+   * that are automatically generated for you, specify the version here.
+   *
+   * You should not typically need to specify this value.
+   *
+   * @default - Latest version
+   */
+  readonly cdkCliVersion?: string;
+
+  /**
+   * Whether the pipeline will update itself
+   *
+   * This needs to be set to `true` to allow the pipeline to reconfigure
+   * itself when assets or stages are being added to it, and `true` is the
+   * recommended setting.
+   *
+   * You can temporarily set this to `false` while you are iterating
+   * on the pipeline itself and prefer to deploy changes using `cdk deploy`.
+   *
+   * @default true
+   */
+  readonly selfMutation?: boolean;
+
+  /**
+   * Set if the pipeline itself builds Docker container assets
+   *
+   * NOTE: this only applies to Docker assets used for the Pipeline
+   * or the Pipeline stack itself. It does not apply for Docker assets
+   * used in the Stages and Stacks that are *deployed* by this pipeline.
+   *
+   * Configures privileged mode for the self-mutation CodeBuild action.
+   *
+   * @default false
+   */
+  readonly pipelineUsesDockerAssets?: boolean;
+
+  /**
+   * Customize the CodeBuild projects created for this pipeline
+   *
+   * @default - All projects run non-privileged build, SMALL instance, LinuxBuildImage.STANDARD_5_0
+   */
+  readonly codeBuildDefaults?: CodeBuildDefaults;
```

Comment:
Why "defaults" instead of "options"?

```suggestion
  readonly codeBuildOptions?: CodeBuildDefaults;
```

Reply:
Because they can/will be combined with other CodeBuild settings afterwards.

Reply:
Will they overridden by other settings later or just merged? If they are overridden, then "defaults" might make sense. If they are merged, then I'd still prefer "options".

Reply:
Merged and overridden both, depending on whether the options are collections or scalars.